collector/compression/zstd/zstd.go

```diff
@@ -0,0 +1,297 @@
+	cfg  EncoderConfig
+	pool mru[*writer]
+}
+
+type decoder struct {
+	lock sync.Mutex // protects cfg
+	cfg  DecoderConfig
+	pool mru[*reader]
+}
+
+type reader struct {
+	*zstdlib.Decoder
+	Gen
+	pool *mru[*reader]
+}
+
+type writer struct {
+	*zstdlib.Encoder
+	Gen
+	pool *mru[*writer]
+}
+
+type combined struct {
+	enc encoder
+	dec decoder
+}
+
+type instance struct {
+	lock    sync.Mutex
+	byLevel map[Level]*combined
+}
+
+var _ encoding.Compressor = &combined{}
+
+var staticInstances = &instance{
+	byLevel: map[Level]*combined{},
+}
+
+func (g *Gen) generation() Gen {
+	return *g
+}
+
+func DefaultEncoderConfig() EncoderConfig {
+	return EncoderConfig{
+		Level:       DefaultLevel, // Determines other defaults
+		Concurrency: 1,            // Avoids extra CPU/memory
+	}
+}
+
+func DefaultDecoderConfig() DecoderConfig {
+	return DecoderConfig{
+		Concurrency:      1,   // Avoids extra CPU/memory
+		MemoryLimitMiB:   128, // More conservative than library default
+		MaxWindowSizeMiB: 32,  // Corresponds w/ "best" level default
+	}
+}
+
+func validate(level Level, f func() error) error {
+	if level > MaxLevel {
+		return fmt.Errorf("level out of range [0,10]: %d", level)
+	}
+	if level < MinLevel {
+		return fmt.Errorf("level out of range [0,10]: %d", level)
+	}
+	return f()
+}
+
+func (cfg EncoderConfig) Validate() error {
+	return validate(cfg.Level, func() error {
+		var buf bytes.Buffer
+		test, err := zstdlib.NewWriter(&buf, cfg.options()...)
+		if test != nil {
+			test.Close()
+		}
+		return err
+	})
+}
+
+func (cfg DecoderConfig) Validate() error {
+	return validate(MinLevel, func() error {
+		var buf bytes.Buffer
+		test, err := zstdlib.NewReader(&buf, cfg.options()...)
+		if test != nil {
+			test.Close()
+		}
+		return err
+	})
+}
+
+func init() {
+	staticInstances.lock.Lock()
+	defer staticInstances.lock.Unlock()
+	resetLibrary()
+}
+
+func resetLibrary() {
+	for level := Level(MinLevel); level <= MaxLevel; level++ {
+		var combi combined
+		combi.enc.cfg = DefaultEncoderConfig()
+		combi.dec.cfg = DefaultDecoderConfig()
+		combi.enc.cfg.Level = level
+		encoding.RegisterCompressor(&combi)
+		staticInstances.byLevel[level] = &combi
+	}
+}
+
+func SetEncoderConfig(cfg EncoderConfig) error {
+	if err := cfg.Validate(); err != nil {
+		return err
+	}
+
+	updateOne := func(enc *encoder) {
+		enc.lock.Lock()
+		defer enc.lock.Unlock()
+		enc.cfg = cfg
+		enc.pool.Reset()
+	}
+
+	staticInstances.lock.Lock()
+	defer staticInstances.lock.Unlock()
+
+	updateOne(&staticInstances.byLevel[cfg.Level].enc)
+	return nil
+}
+
+func SetDecoderConfig(cfg DecoderConfig) error {
+	if err := cfg.Validate(); err != nil {
+		return err
+	}
+	updateOne := func(dec *decoder) {
+		dec.lock.Lock()
+		defer dec.lock.Unlock()
+		dec.cfg = cfg
+		dec.pool.Reset()
+	}
+
+	staticInstances.lock.Lock()
+	defer staticInstances.lock.Unlock()
+
+	for level := MinLevel; level <= MaxLevel; level++ {
+		updateOne(&staticInstances.byLevel[level].dec)
```

Comment:
Fine question! I have gone both ways on this question during development. Firstly, encoders have levels, but decoders only have limits. 

As structured, the decoder level used in this code is an artifact of allowing the encoder to configure levels, because the name of the PAIR changes when the encoder level changes. 

Given that the notion of decoder level serves only to match the decoder limits with encoder limits, then it's worth asking whether it would be useful. Imagine configuring permissive limits for high-level compression and smaller limits for lower compression levels. This makes sense, but it might not protect the user against in a hostile environment. The goal of these limits is to protect the user, so having one limit is just easier.

I originally coded multiple limits, then reasoned through what the user would see. The OTel-Arrow receiver component would be able to configure different limits for every level, so would have a slice of decoder configs. This would make it hard to simply configure a limit, because you'd have to configure them all. 

The code could be restructured to not have 10 distinct decoders, or it could be extended back in the other direction, to allow distinct decoder configurations. For now, for simplicity, I chose this middle ground.